File: Medium/785. Is Graph Bipartite/Solution.py
ans = True

def dfs(graph, kind, vis, node):
    global ans
    vis[node] = True

    for child in graph[node]:
        if not vis[child]:
            kind[child] = 1 - kind[node]
            dfs(graph, kind, vis, child)
        elif kind[child] == kind[node]:
            ans = False

class Solution(object):
    def isBipartite(self, graph):
        global ans
        n = len(graph)
        kind = [0] * n
        vis = [False] * n
        ans = True

        for i in range(n):
            if not vis[i]:
                dfs(graph, kind, vis, i)

        return ans

File: Medium/785. Is Graph Bipartite/Solution.cpp
class Solution {
public:
    void dfs(vector<vector<int> > & graph, vector<int> & kind, vector<bool> &vis, int node, bool & bipartite) {
        vis[node] = true;

        for(auto & child : graph[node]) {
            if(!vis[child]) {
                kind[child] = 1 - kind[node];
                dfs(graph, kind, vis, child, bipartite);
            }
            else if(kind[child] == kind[node])
                bipartite = false;
        }
    }
    bool isBipartite(vector<vector<int>>& graph) {
        int n = (int)graph.size();
        vector<int> kind(n , 0);
        vector<bool> vis(n, 0);
        bool bipartite = true;

        for(int i = 0; i < n; i++)
            if(!vis[i])
                dfs(graph, kind, vis, i, bipartite);

        return bipartite;
    }
};